diff --git a/lib/typescript/hooks/useBottomSheetDynamicSnapPoints.d.ts b/lib/typescript/hooks/useBottomSheetDynamicSnapPoints.d.ts
index 36db692d081b037cbe450ca31d186f8cc4c8410f..12d62711d3214f264bfb8bd4ce586f189dd5cece 100644
--- a/lib/typescript/hooks/useBottomSheetDynamicSnapPoints.d.ts
+++ b/lib/typescript/hooks/useBottomSheetDynamicSnapPoints.d.ts
@@ -14,9 +14,7 @@
  * @deprecated will be deprecated in the next major release! please use the new introduce prop `enableDynamicSizing`.
  */
 export declare const useBottomSheetDynamicSnapPoints: (initialSnapPoints: Array<string | number>) => {
-    animatedSnapPoints: Readonly<{
-        value: (string | number)[];
-    }>;
+    animatedSnapPoints: import("react-native-reanimated").SharedValue<(string | number)[]>;
     animatedHandleHeight: import("react-native-reanimated").SharedValue<number>;
     animatedContentHeight: import("react-native-reanimated").SharedValue<number>;
     handleContentLayout: ({ nativeEvent: { layout: { height }, }, }: {
