diff --git a/registry.js b/registry.js
index 64b2735d3bb5284bd2450bf0d06115c3de5dcf80..4121e73b39a1c5d72144dcc98c1187641ab7a4f8 100644
--- a/registry.js
+++ b/registry.js
@@ -8,8 +8,9 @@
  * @format
  */
 
-'use strict';
+"use strict";
 
+/*::
 export type AssetDestPathResolver = 'android' | 'generic';
 
 export type PackagerAsset = {
@@ -25,17 +26,17 @@ export type PackagerAsset = {
   +resolver?: AssetDestPathResolver,
   ...
 };
+*/
+const assets /*::: Array<PackagerAsset>*/ = [];
 
-const assets: Array<PackagerAsset> = [];
-
-function registerAsset(asset: PackagerAsset): number {
+function registerAsset(asset /*::: PackagerAsset*/) /*::: number*/ {
   // `push` returns new array length, so the first asset will
   // get id 1 (not 0) to make the value truthy
   return assets.push(asset);
 }
 
-function getAssetByID(assetId: number): PackagerAsset {
+function getAssetByID(assetId /*::: number*/) /*::: PackagerAsset*/ {
   return assets[assetId - 1];
 }
 
-module.exports = {registerAsset, getAssetByID};
+module.exports = { registerAsset, getAssetByID };
