diff --git a/settings-plugin/src/main/kotlin/com/facebook/react/ReactSettingsExtension.kt b/settings-plugin/src/main/kotlin/com/facebook/react/ReactSettingsExtension.kt
index aea525747e08b811eabae78c4be486fe7b3c46ba..a70b9cd72346487d347337cc95c24a4a3d8652f8 100644
--- a/settings-plugin/src/main/kotlin/com/facebook/react/ReactSettingsExtension.kt
+++ b/settings-plugin/src/main/kotlin/com/facebook/react/ReactSettingsExtension.kt
@@ -29,7 +29,7 @@ abstract class ReactSettingsExtension @Inject constructor(val settings: Settings
       settings.layout.rootDirectory.file("build/generated/autolinking/").asFile
 
   private val defaultConfigCommand: List<String> =
-      windowsAwareCommandLine(listOf("npx", "@react-native-community/cli", "config")).map {
+      windowsAwareCommandLine(listOf("node", "../../node_modules/@react-native-community/cli/build/bin.js", "config")).map {
         it.toString()
       }
 
