diff --git a/dist/index.js b/dist/index.js
index 015a33c37fe87f13f31559d462351acd7ae9bac7..4cd7cdeb4437f30c1c063c0ffb8fd5692a399dbf 100644
--- a/dist/index.js
+++ b/dist/index.js
@@ -68,8 +68,8 @@ class GnosisSafe extends types_1.Connector {
             if (this.eagerConnection)
                 return;
             // kick off import early to minimize waterfalls
-            const SafeAppProviderPromise = Promise.resolve().then(() => __importStar(require('@safe-global/safe-apps-provider'))).then(({ SafeAppProvider }) => SafeAppProvider);
-            yield (this.eagerConnection = Promise.resolve().then(() => __importStar(require('@safe-global/safe-apps-sdk'))).then((m) => __awaiter(this, void 0, void 0, function* () {
+            const SafeAppProviderPromise = Promise.resolve().then(async () => __importStar(await import('@safe-global/safe-apps-provider'))).then(({ SafeAppProvider }) => SafeAppProvider);
+            yield (this.eagerConnection = Promise.resolve().then(async () => __importStar(await import('@safe-global/safe-apps-sdk'))).then((m) => __awaiter(this, void 0, void 0, function* () {
                 this.sdk = new m.default(this.options);
                 const safe = yield Promise.race([
                     this.sdk.safe.getInfo(),
