diff --git a/config/jest/fileTransform.js b/config/jest/fileTransform.js
index aab67618c38b46485f28d18bd81f7125383037ed..95c34bb742cf15e8da436d412bcbf623c8185e0b 100644
--- a/config/jest/fileTransform.js
+++ b/config/jest/fileTransform.js
@@ -1,7 +1,7 @@
-'use strict';
+"use strict";
 
-const path = require('path');
-const camelcase = require('camelcase');
+const path = require("path");
+const camelcase = require("camelcase");
 
 // This is a custom Jest transformer turning file imports into filenames.
 // http://facebook.github.io/jest/docs/en/webpack.html
@@ -17,7 +17,8 @@ module.exports = {
         pascalCase: true,
       });
       const componentName = `Svg${pascalCaseFilename}`;
-      return `const React = require('react');
+      return {
+        code: `const React = require('react');
       module.exports = {
         __esModule: true,
         default: ${assetFilename},
@@ -32,9 +33,12 @@ module.exports = {
             })
           };
         }),
-      };`;
+      };`,
+      };
     }
 
-    return `module.exports = ${assetFilename};`;
+    return {
+      code: `module.exports = ${assetFilename};`,
+    };
   },
 };
