diff --git a/react-native-slider.podspec b/react-native-slider.podspec
--- a/react-native-slider.podspec
+++ b/react-native-slider.podspec
@@ -13,7 +13,13 @@ Pod::Spec.new do |s|
   s.platforms    = { :ios => "9.0", :visionos => "1.0" }

   s.source       = { :git => "https://github.com/callstack/react-native-slider.git", :tag => "v#{s.version}" }
-  s.source_files = "ios/**/*.{h,m,mm}"
+
+  if ENV['RCT_NEW_ARCH_ENABLED'] == '1'
+    s.source_files = "ios/**/*.{h,m,mm}"
+  else
+    s.source_files = "ios/**/*.{h,m}"
+    s.exclude_files = "ios/**/RNCSliderComponentView.{h,mm}"
+  end

   s.subspec "common" do |ss|
     ss.source_files         = "common/cpp/**/*.{cpp,h}"
