diff --git a/node_modules/@web3-react/walletconnect-v2/dist/index.js b/node_modules/@web3-react/walletconnect-v2/dist/index.js
index 1a36d14..908b8c5 100644
--- a/node_modules/@web3-react/walletconnect-v2/dist/index.js
+++ b/node_modules/@web3-react/walletconnect-v2/dist/index.js
@@ -84,7 +84,7 @@ class WalletConnect extends types_1.Connector {
         return __awaiter(this, void 0, void 0, function* () {
             const rpcMap = this.rpcMap ? (0, utils_1.getBestUrlMap)(this.rpcMap, this.timeout) : undefined;
             const chainProps = this.getChainProps(this.chains, this.optionalChains, desiredChainId);
-            const ethProviderModule = yield Promise.resolve().then(() => __importStar(require('@walletconnect/ethereum-provider')));
+            const ethProviderModule = yield Promise.resolve().then(async () => __importStar(await import('@walletconnect/ethereum-provider')));
             this.provider = yield ethProviderModule.default.init(Object.assign(Object.assign(Object.assign({}, this.options), chainProps), { rpcMap: yield rpcMap }));
             return this.provider
                 .on('disconnect', this.disconnectListener)
diff --git a/node_modules/@web3-react/walletconnect-v2/dist/utils.js b/node_modules/@web3-react/walletconnect-v2/dist/utils.js
index 17539b6..9ea6371 100644
--- a/node_modules/@web3-react/walletconnect-v2/dist/utils.js
+++ b/node_modules/@web3-react/walletconnect-v2/dist/utils.js
@@ -62,8 +62,8 @@ function getBestUrl(urls, timeout) {
         if (urls.length === 1)
             return urls[0];
         const [HttpConnection, JsonRpcProvider] = yield Promise.all([
-            Promise.resolve().then(() => __importStar(require('@walletconnect/jsonrpc-http-connection'))).then(({ HttpConnection }) => HttpConnection),
-            Promise.resolve().then(() => __importStar(require('@walletconnect/jsonrpc-provider'))).then(({ JsonRpcProvider }) => JsonRpcProvider),
+            Promise.resolve().then(async () => __importStar(await import('@walletconnect/jsonrpc-http-connection'))).then(({ HttpConnection }) => HttpConnection),
+            Promise.resolve().then(async () => __importStar(await import('@walletconnect/jsonrpc-provider'))).then(({ JsonRpcProvider }) => JsonRpcProvider),
         ]);
         // the below returns the first url for which there's been a successful call, prioritized by index
         return new Promise((resolve) => {
