diff --git a/node_modules/@web3-react/coinbase-wallet/dist/index.js b/node_modules/@web3-react/coinbase-wallet/dist/index.js
index f38d06e..4f8fa19 100644
--- a/node_modules/@web3-react/coinbase-wallet/dist/index.js
+++ b/node_modules/@web3-react/coinbase-wallet/dist/index.js
@@ -62,7 +62,7 @@ class CoinbaseWallet extends types_1.Connector {
         return __awaiter(this, void 0, void 0, function* () {
             if (this.eagerConnection)
                 return;
-            yield (this.eagerConnection = Promise.resolve().then(() => __importStar(require('@coinbase/wallet-sdk'))).then((m) => {
+            yield (this.eagerConnection = Promise.resolve().then(async () => __importStar(await import('@coinbase/wallet-sdk'))).then((m) => {
                 const _a = this.options, { url } = _a, options = __rest(_a, ["url"]);
                 this.coinbaseWallet = new m.default(options);
                 this.provider = this.coinbaseWallet.makeWeb3Provider(url);
