diff --git a/node_modules/colorthief/src/color-thief-node.d.ts b/node_modules/colorthief/src/color-thief-node.d.ts
new file mode 100644
index 0000000..af733cf
--- /dev/null
+++ b/node_modules/colorthief/src/color-thief-node.d.ts
@@ -0,0 +1,2 @@
+declare module 'colorthief/src/color-thief-node'
+
diff --git a/node_modules/colorthief/src/color-thief-node.js b/node_modules/colorthief/src/color-thief-node.js
index efc73b3..3c8317f 100644
--- a/node_modules/colorthief/src/color-thief-node.js
+++ b/node_modules/colorthief/src/color-thief-node.js
@@ -1,4 +1,4 @@
-const getPixels = require('get-pixels');
+var PNG = require('pngjs').PNG
 const quantize = require('@lokesh.dhakar/quantize');
 
 function createPixelArray(imgData, pixelCount, quality) {
@@ -45,15 +45,23 @@ function validateOptions(options) {
 }
 
 function loadImg(img) {
-    return new Promise((resolve, reject) => {
-        getPixels(img, function(err, data) {
+    const p1 = new Promise((resolve, reject) => {
+        var png = new PNG();
+        png.parse(img, function(err, img_data) {
             if(err) {
                 reject(err)
             } else {
-                resolve(data);
+                resolve(img_data);
             }
         })
     });
+    const p2 = new Promise((resolve, reject) => {
+        setTimeout(() => {
+            reject('timeout');
+        }
+        , 10);
+    });
+    return Promise.race([p1, p2]);
 }
 
 function getColor(img, quality) {
@@ -75,10 +83,11 @@ function getPalette(img, colorCount = 10, quality = 10) {
         quality
     });
     
+
     return new Promise((resolve, reject) => {
         loadImg(img)
             .then(imgData => {
-                const pixelCount = imgData.shape[0] * imgData.shape[1];
+                const pixelCount = imgData.height * imgData.width;
                 const pixelArray = createPixelArray(imgData.data, pixelCount, options.quality);
 
                 const cmap = quantize(pixelArray, options.colorCount);
