diff --git a/node_modules/colorthief/src/color-thief-node.js b/node_modules/colorthief/src/color-thief-node.js
index efc73b3..74b5384 100644
--- a/node_modules/colorthief/src/color-thief-node.js
+++ b/node_modules/colorthief/src/color-thief-node.js
@@ -1,4 +1,4 @@
-const getPixels = require('get-pixels');
+var PNG = require('pngjs').PNG
 const quantize = require('@lokesh.dhakar/quantize');
 
 function createPixelArray(imgData, pixelCount, quality) {
@@ -46,11 +46,12 @@ function validateOptions(options) {
 
 function loadImg(img) {
     return new Promise((resolve, reject) => {
-        getPixels(img, function(err, data) {
+        var png = new PNG();
+        png.parse(img, function(err, img_data) {
             if(err) {
                 reject(err)
             } else {
-                resolve(data);
+                resolve(img_data);
             }
         })
     });
@@ -78,7 +79,7 @@ function getPalette(img, colorCount = 10, quality = 10) {
     return new Promise((resolve, reject) => {
         loadImg(img)
             .then(imgData => {
-                const pixelCount = imgData.shape[0] * imgData.shape[1];
+                const pixelCount = imgData.height * imgData.width;
                 const pixelArray = createPixelArray(imgData.data, pixelCount, options.quality);
 
                 const cmap = quantize(pixelArray, options.colorCount);
