diff --git a/node_modules/ext-name/index.js b/node_modules/ext-name/index.js
index 59ac1f6..fa857ef 100644
--- a/node_modules/ext-name/index.js
+++ b/node_modules/ext-name/index.js
@@ -4,7 +4,7 @@ const sortKeysLength = require('sort-keys-length');
 
 module.exports = str => {
 	const obj = sortKeysLength.desc(extList());
-	const exts = Object.keys(obj).filter(x => str.endsWith(x));
+	const exts = Object.keys(obj).filter(x => str.indexOf(x, str.length - x.length) !== -1);
 
 	if (exts.length === 0) {
 		return [];
