diff --git a/node_modules/get-pixels/dom-pixels.js b/node_modules/get-pixels/dom-pixels.js
index 7714528..64e8db3 100644
--- a/node_modules/get-pixels/dom-pixels.js
+++ b/node_modules/get-pixels/dom-pixels.js
@@ -1,10 +1,8 @@
 'use strict'
 
-var path          = require('path')
+var extname       = require('ext-name')
 var ndarray       = require('ndarray')
 var GifReader     = require('omggif').GifReader
-var pack          = require('ndarray-pack')
-var through       = require('through')
 var parseDataURI  = require('data-uri-to-buffer')
 
 function defaultImage(url, cb) {
@@ -117,9 +115,9 @@ module.exports = function getPixels(url, type, cb) {
     cb = type
     type = ''
   }
-  var ext = path.extname(url)
+  var ext = extname(url).ext
   switch(type || ext.toUpperCase()) {
-    case '.GIF':
+    case 'GIF':
       httpGif(url, cb)
     break
     default:
