diff --git a/node_modules/get-pixels/dom-pixels.js b/node_modules/get-pixels/dom-pixels.js
index 7714528..74e01fe 100644
--- a/node_modules/get-pixels/dom-pixels.js
+++ b/node_modules/get-pixels/dom-pixels.js
@@ -1,11 +1,10 @@
 'use strict'
 
-var path          = require('path')
+var extname       = require('ext-name')
 var ndarray       = require('ndarray')
 var GifReader     = require('omggif').GifReader
-var pack          = require('ndarray-pack')
-var through       = require('through')
 var parseDataURI  = require('data-uri-to-buffer')
+var Buffer        = require('buffer').Buffer
 
 function defaultImage(url, cb) {
   var img = new Image()
@@ -117,12 +116,24 @@ module.exports = function getPixels(url, type, cb) {
     cb = type
     type = ''
   }
-  var ext = path.extname(url)
-  switch(type || ext.toUpperCase()) {
-    case '.GIF':
-      httpGif(url, cb)
-    break
-    default:
+  var ext = extname(url).ext
+  try {
+    switch(type || ext.toUpperCase()) {
+      case 'GIF':
+        httpGif(url, cb)
+      break
+      default:
+        if(Buffer.isBuffer(url)) {
+          url = 'data:' + type + ';base64,' + url.toString('base64')
+        }
+        if(url.indexOf('data:image/gif;') === 0) {
+          dataGif(url, cb)
+        } else {
+          defaultImage(url, cb)
+        }
+    }
+  } catch(err) {
+    try {
       if(Buffer.isBuffer(url)) {
         url = 'data:' + type + ';base64,' + url.toString('base64')
       }
@@ -131,5 +142,8 @@ module.exports = function getPixels(url, type, cb) {
       } else {
         defaultImage(url, cb)
       }
+    } catch(err2) {
+      cb(err2)
+    }
   }
 }
\ No newline at end of file
