// wagmi/chains/dist/wagmi-chains.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var chains = require('@wagmi/core/chains');



Object.defineProperty(exports, 'arbitrum', {
  enumerable: true,
  get: function () { return chains.arbitrum; }
});
Object.defineProperty(exports, 'arbitrumRinkeby', {
  enumerable: true,
  get: function () { return chains.arbitrumRinkeby; }
});
Object.defineProperty(exports, 'foundry', {
  enumerable: true,
  get: function () { return chains.foundry; }
});
Object.defineProperty(exports, 'goerli', {
  enumerable: true,
  get: function () { return chains.goerli; }
});
Object.defineProperty(exports, 'hardhat', {
  enumerable: true,
  get: function () { return chains.hardhat; }
});
Object.defineProperty(exports, 'kovan', {
  enumerable: true,
  get: function () { return chains.kovan; }
});
Object.defineProperty(exports, 'localhost', {
  enumerable: true,
  get: function () { return chains.localhost; }
});
Object.defineProperty(exports, 'mainnet', {
  enumerable: true,
  get: function () { return chains.mainnet; }
});
Object.defineProperty(exports, 'optimism', {
  enumerable: true,
  get: function () { return chains.optimism; }
});
Object.defineProperty(exports, 'optimismKovan', {
  enumerable: true,
  get: function () { return chains.optimismKovan; }
});
Object.defineProperty(exports, 'polygon', {
  enumerable: true,
  get: function () { return chains.polygon; }
});
Object.defineProperty(exports, 'polygonMumbai', {
  enumerable: true,
  get: function () { return chains.polygonMumbai; }
});
Object.defineProperty(exports, 'rinkeby', {
  enumerable: true,
  get: function () { return chains.rinkeby; }
});
Object.defineProperty(exports, 'ropsten', {
  enumerable: true,
  get: function () { return chains.ropsten; }
});
