// tsdx/dist/babelPluginTsdx.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@babel/core");
const rollup_plugin_babel_1 = tslib_1.__importDefault(require("rollup-plugin-babel"));
const lodash_merge_1 = tslib_1.__importDefault(require("lodash.merge"));
exports.isTruthy = (obj) => {
    if (!obj) {
        return false;
    }
    return obj.constructor !== Object || Object.keys(obj).length > 0;
};
const replacements = [{ original: 'lodash', replacement: 'lodash-es' }];
exports.mergeConfigItems = (type, ...configItemsToMerge) => {
    const mergedItems = [];
    configItemsToMerge.forEach(configItemToMerge => {
        configItemToMerge.forEach((item) => {
            const itemToMergeWithIndex = mergedItems.findIndex(mergedItem => mergedItem.file.resolved === item.file.resolved);
            if (itemToMergeWithIndex === -1) {
                mergedItems.push(item);
                return;
            }
            mergedItems[itemToMergeWithIndex] = core_1.createConfigItem([
                mergedItems[itemToMergeWithIndex].file.resolved,
                lodash_merge_1.default(mergedItems[itemToMergeWithIndex].options, item.options),
            ], {
                type,
            });
        });
    });
    return mergedItems;
};
exports.createConfigItems = (type, items) => {
    return items.map((_a) => {
        var { name } = _a, options = tslib_1.__rest(_a, ["name"]);
        return core_1.createConfigItem([require.resolve(name), options], { type });
    });
};
exports.babelPluginTsdx = rollup_plugin_babel_1.default.custom((babelCore) => ({
    // Passed the plugin options.
    options(_a) {
        var { custom: customOptions } = _a, pluginOptions = tslib_1.__rest(_a, ["custom"]);
        return {
            // Pull out any custom options that the plugin might have.
            customOptions,
            // Pass the options back with the two custom options removed.
            pluginOptions,
        };
    },
    config(config, { customOptions }) {
        const defaultPlugins = exports.createConfigItems('plugin', [
            // {
            //   name: '@babel/plugin-transform-react-jsx',
            //   pragma: customOptions.jsx || 'h',
            //   pragmaFrag: customOptions.jsxFragment || 'Fragment',
            // },
            { name: 'babel-plugin-annotate-pure-calls' },
            { name: 'babel-plugin-dev-expression' },
            customOptions.format !== 'cjs' && {
                name: 'babel-plugin-transform-rename-import',
                replacements,
            },
            exports.isTruthy(customOptions.defines) && {
                name: 'babel-plugin-transform-replace-expressions',
                replace: customOptions.defines,
            },
            {
                name: 'babel-plugin-transform-async-to-promises',
                inlineHelpers: true,
                externalHelpers: true,
            },
            {
                name: '@babel/plugin-proposal-class-properties',
                loose: true,
            },
            // Adds syntax support for optional chaining (.?)
            { name: '@babel/plugin-proposal-optional-chaining' },
            // Adds syntax support for default value using ?? operator
            { name: '@babel/plugin-proposal-nullish-coalescing-operator' },
            {
                name: '@babel/plugin-transform-regenerator',
                async: false,
            },
            {
                name: 'babel-plugin-macros',
            },
            exports.isTruthy(customOptions.extractErrors) && {
                name: './errors/transformErrorMessages',
            },
        ].filter(Boolean));
        const babelOptions = config.options || {};
        const envIdx = (babelOptions.presets || []).findIndex((preset) => preset.file.request.includes('@babel/preset-env'));
        if (envIdx !== -1) {
            const preset = babelOptions.presets[envIdx];
            babelOptions.presets[envIdx] = core_1.createConfigItem([
                preset.file.resolved,
                lodash_merge_1.default({
                    loose: true,
                    targets: customOptions.targets,
                }, preset.options, {
                    modules: false,
                    exclude: lodash_merge_1.default(['transform-async-to-generator', 'transform-regenerator'], (preset.options && preset.options.exclude) || []),
                }),
            ], {
                type: `preset`,
            });
        }
        else {
            babelOptions.presets = exports.createConfigItems('preset', [
                {
                    name: '@babel/preset-env',
                    targets: customOptions.targets,
                    modules: false,
                    loose: true,
                    exclude: ['transform-async-to-generator', 'transform-regenerator'],
                },
            ]);
        }
        // Merge babelrc & our plugins together
        babelOptions.plugins = exports.mergeConfigItems('plugin', defaultPlugins, babelOptions.plugins || []);
        return babelOptions;
    },
}));
