// wagmi/actions/dist/wagmi-actions.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@wagmi/core');



Object.defineProperty(exports, 'connect', {
  enumerable: true,
  get: function () { return core.connect; }
});
Object.defineProperty(exports, 'disconnect', {
  enumerable: true,
  get: function () { return core.disconnect; }
});
Object.defineProperty(exports, 'fetchBalance', {
  enumerable: true,
  get: function () { return core.fetchBalance; }
});
Object.defineProperty(exports, 'fetchBlockNumber', {
  enumerable: true,
  get: function () { return core.fetchBlockNumber; }
});
Object.defineProperty(exports, 'fetchEnsAddress', {
  enumerable: true,
  get: function () { return core.fetchEnsAddress; }
});
Object.defineProperty(exports, 'fetchEnsAvatar', {
  enumerable: true,
  get: function () { return core.fetchEnsAvatar; }
});
Object.defineProperty(exports, 'fetchEnsName', {
  enumerable: true,
  get: function () { return core.fetchEnsName; }
});
Object.defineProperty(exports, 'fetchEnsResolver', {
  enumerable: true,
  get: function () { return core.fetchEnsResolver; }
});
Object.defineProperty(exports, 'fetchFeeData', {
  enumerable: true,
  get: function () { return core.fetchFeeData; }
});
Object.defineProperty(exports, 'fetchSigner', {
  enumerable: true,
  get: function () { return core.fetchSigner; }
});
Object.defineProperty(exports, 'fetchToken', {
  enumerable: true,
  get: function () { return core.fetchToken; }
});
Object.defineProperty(exports, 'getAccount', {
  enumerable: true,
  get: function () { return core.getAccount; }
});
Object.defineProperty(exports, 'getContract', {
  enumerable: true,
  get: function () { return core.getContract; }
});
Object.defineProperty(exports, 'getNetwork', {
  enumerable: true,
  get: function () { return core.getNetwork; }
});
Object.defineProperty(exports, 'getProvider', {
  enumerable: true,
  get: function () { return core.getProvider; }
});
Object.defineProperty(exports, 'getWebSocketProvider', {
  enumerable: true,
  get: function () { return core.getWebSocketProvider; }
});
Object.defineProperty(exports, 'readContract', {
  enumerable: true,
  get: function () { return core.readContract; }
});
Object.defineProperty(exports, 'readContracts', {
  enumerable: true,
  get: function () { return core.readContracts; }
});
Object.defineProperty(exports, 'sendTransaction', {
  enumerable: true,
  get: function () { return core.sendTransaction; }
});
Object.defineProperty(exports, 'signMessage', {
  enumerable: true,
  get: function () { return core.signMessage; }
});
Object.defineProperty(exports, 'signTypedData', {
  enumerable: true,
  get: function () { return core.signTypedData; }
});
Object.defineProperty(exports, 'switchNetwork', {
  enumerable: true,
  get: function () { return core.switchNetwork; }
});
Object.defineProperty(exports, 'waitForTransaction', {
  enumerable: true,
  get: function () { return core.waitForTransaction; }
});
Object.defineProperty(exports, 'watchAccount', {
  enumerable: true,
  get: function () { return core.watchAccount; }
});
Object.defineProperty(exports, 'watchBlockNumber', {
  enumerable: true,
  get: function () { return core.watchBlockNumber; }
});
Object.defineProperty(exports, 'watchContractEvent', {
  enumerable: true,
  get: function () { return core.watchContractEvent; }
});
Object.defineProperty(exports, 'watchNetwork', {
  enumerable: true,
  get: function () { return core.watchNetwork; }
});
Object.defineProperty(exports, 'watchProvider', {
  enumerable: true,
  get: function () { return core.watchProvider; }
});
Object.defineProperty(exports, 'watchReadContract', {
  enumerable: true,
  get: function () { return core.watchReadContract; }
});
Object.defineProperty(exports, 'watchReadContracts', {
  enumerable: true,
  get: function () { return core.watchReadContracts; }
});
Object.defineProperty(exports, 'watchSigner', {
  enumerable: true,
  get: function () { return core.watchSigner; }
});
Object.defineProperty(exports, 'watchWebSocketProvider', {
  enumerable: true,
  get: function () { return core.watchWebSocketProvider; }
});
Object.defineProperty(exports, 'writeContract', {
  enumerable: true,
  get: function () { return core.writeContract; }
});
