// react-query/lib/createAsyncStoragePersister/index.js
"use strict";

exports.__esModule = true;
exports.createAsyncStoragePersister = void 0;

var _asyncThrottle = require("./asyncThrottle");

var _utils = require("../core/utils");

const createAsyncStoragePersister = ({
  storage,
  key = "REACT_QUERY_OFFLINE_CACHE",
  throttleTime = 1000,
  serialize = JSON.stringify,
  deserialize = JSON.parse,
  retry
}) => {
  if (typeof storage !== 'undefined') {
    const trySave = async persistedClient => {
      try {
        await storage.setItem(key, serialize(persistedClient));
      } catch (error) {
        return error;
      }
    };

    return {
      persistClient: (0, _asyncThrottle.asyncThrottle)(async persistedClient => {
        let client = persistedClient;
        let error = await trySave(client);
        let errorCount = 0;

        while (error && client) {
          errorCount++;
          client = await (retry == null ? void 0 : retry({
            persistedClient: client,
            error,
            errorCount
          }));

          if (client) {
            error = await trySave(client);
          }
        }
      }, {
        interval: throttleTime
      }),
      restoreClient: async () => {
        const cacheString = await storage.getItem(key);

        if (!cacheString) {
          return;
        }

        return deserialize(cacheString);
      },
      removeClient: () => storage.removeItem(key)
    };
  }

  return {
    persistClient: _utils.noop,
    restoreClient: _utils.noop,
    removeClient: _utils.noop
  };
};

exports.createAsyncStoragePersister = createAsyncStoragePersister;
